src/config.cpp

```diff
@@ -0,0 +1,322 @@
+
+    for (auto it = manifest.begin(); it != manifest.end(); ++it) {
+        const auto name = it.key();
+
+        if (std::regex_match(name, regexVariable)) {
+            variables[name] = it.value();
+        } else {
+            throw SonataError(fmt::format("Invalid variable `{}`", name));
+        }
+    }
+
+    return variables;
+}
+
+nlohmann::json parseSonataJson(const std::string& contents) {
+    const auto json = nlohmann::json::parse(contents);
+
+    const auto vars = replaceVariables(readVariables(json));
+    return expandVariables(json, vars);
+}
+
+
+template <typename Population>
+std::map<std::string, SubnetworkFiles> resolvePopulations(
+    const std::vector<SubnetworkFiles>& networkNodes) {
+    using PopulationStorage = bbp::sonata::PopulationStorage<Population>;
+    std::map<std::string, SubnetworkFiles> result;
+
+    for (const auto& network : networkNodes) {
+        const auto population = PopulationStorage(network.elements, network.types);
+        for (auto name : population.populationNames()) {
+            result[name] = network;
+        }
+    }
+    return result;
+}
+
+template <typename Population>
+std::set<std::string> listPopulations(const std::vector<SubnetworkFiles>& networkNodes) {
+    std::set<std::string> result;
+    const auto names = resolvePopulations<Population>(networkNodes);
+    std::transform(names.begin(),
+                   names.end(),
+                   std::inserter(result, result.end()),
+                   [](std::pair<std::string, SubnetworkFiles> p) { return p.first; });
+    return result;
+}
+}  // namespace
+
+
+namespace bbp {
+namespace sonata {
+
+struct CircuitConfig::Impl {
+    PathResolver resolver;
+
+    std::string target_simulator;
+    std::string node_sets_file;
+    std::map<std::string, std::string> components;
+    std::vector<SubnetworkFiles> networkNodes;
+    std::vector<SubnetworkFiles> networkEdges;
+
+    Impl(const std::string& contents, const std::string& basePath)
+        : resolver(basePath) {
+        const auto json = parseSonataJson(contents);
+
+        try {
+            target_simulator = json.at("target_simulator");
+        } catch (nlohmann::detail::out_of_range&) {
+        }
+
+        if (json.find("node_sets_file") != json.end()) {
+            node_sets_file = resolver.toAbsolute(json["node_sets_file"]);
+        }
+
+        if (json.find("networks") == json.end()) {
+            throw SonataError("Error parsing config: `networks` not specified");
+        }
+
+        auto networks = json.at("networks");
+        components = fillComponents(json, resolver);
+        networkNodes = fillSubnetwork(networks, "node", resolver);
+        networkEdges = fillSubnetwork(networks, "edge", resolver);
+    }
+};
+
+CircuitConfig::CircuitConfig(const std::string& contents, const std::string& basePath)
+    : impl(new CircuitConfig::Impl(contents, basePath)) {}
+
+CircuitConfig::CircuitConfig(CircuitConfig&&) = default;
```

Comment:
The compiler should generate those, none of the 5 (from rule of 5) are custom.

Reply:
Because the copy constructor is deleted, in the static initializer `fromFile` the compiler tries to call the copy constructor if the move constructor is not manually defined (not sure why, since as you said, it should be automatically generated). 
They also both need to be defined in the `.cpp` because `CircuitConfig::Impl` is an incomplete type in the header.

Reply:
> They also both need to be defined in the .cpp because CircuitConfig::Impl is an incomplete type in the header.

Yeah; good old c++.  The other option is to drop the `PIMPL`; was just matching the Brion version - up to you @NadirRoGue what you prefer.

I probably should have something like:
```
CircuitConfig& operator=(CircuitConfig&& rhs) = default;
CircuitConfig& operator=(CircuitConfig& rhs) = delete;
```
To match the `CircuitConfig::CircuitConfig(CircuitConfig&&) = default;` and deleted version.

Reply:
I'd rather remove pimpl if everybody is ok with that.

Reply:
Actually I wondered why it was used, If not too laborious, yeah, would go with that